### PR TITLE
Fix DynamoDB getAllRecords logic when 1MB query limit is reached

### PR DESCRIPTION
**Pull request by smallinsky**

### CODE REVIEW

**Review by greedy52 — COMMENTED** [review]
Fix looks good. I think an alternative is to keep a running limit like
```
for i := 0; i < backend.DefaultRangeLimit/100; i++ {
	limitRemaining := 0
	if limit > 0 {
		limitRemaining = limit - len(results.records)
	}
	re, err := b.getRecords(ctx, prependPrefix(startKey), prependPrefix(endKey), limitRemaining, result.lastEvaluatedKey)

...
}
```
This can make sure we WONT fetch more than what we need. Also maybe no need to do `len(result.records) >= limit` check i believe.

File: lib/backend/test/suite.go
Comment (greedy52) [inline lib/backend/test/suite.go:589]: `1<16` is 65K and `65K * 20 > 1MB`?

File: lib/backend/test/suite.go
Comment (greedy52) [inline lib/backend/test/suite.go:573]: curious why `require.Eventually` to replace `clock.Advance`?

File: lib/backend/dynamo/dynamodbbk.go
Comment (greedy52) [inline lib/backend/dynamo/dynamodbbk.go:393]: Really really sorry for going back and forth on this. Now i am a little concerned how we are using the limit.

by [AWS documentation](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html), 
> [Limit](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html#API_Query_RequestSyntax)
The maximum number of items to evaluate (not necessarily the number of matching items).

Meaning the `Limit` is maximum number of items to be scanned, *pre-filtering*. The records we obtained is *after filtering*.


---
The only filtering we have in `getRecords` is `Expires >= :timestamp`. 

Let's use an example, say we have 10 records, and all first 9 records are expired.

If we use `Limit = 5` for this query, we wound **NOT** get any results back (since we scanned first 5, all expired). 

Even though, we have one matching record, we won't get it unless `Limit` is greater than 9. (or no limit)

---
So, in our case, is the `limit` parameter the **limit-of-items-to-scan-for-all-queries** or **limit-of-items-to-scan-per-query** or **limit-of-items-to-return**?

File: lib/backend/test/suite.go
Comment (smallinsky) [reply lib/backend/test/suite.go]: Right, The comment should say 65KB

File: lib/backend/test/suite.go
Comment (smallinsky) [reply lib/backend/test/suite.go:573]: The 3 tests in DynamoDB test suite were failing. The `clock.Advance` doesn't affect the DynamoDB clock so enforcing items to expire by calling `clock.Advance` doesn't work.  So in this case we need to relay on real clock.

Comment (greedy52) [reply lib/backend/test/suite.go:573]: cool. 

for the `Get` failure at least, seems the `record` is our own type. Maybe we can pass the clock to `isExpired` function or assign the clock to the object
https://github.com/gravitational/teleport/blob/cda00f3e798db6582ee2885da96900ed5490a381/lib/backend/dynamo/dynamodbbk.go#L866-L871

I am ok with `Eventually` too. was just curious.

File: lib/backend/dynamo/dynamodbbk.go
Comment (smallinsky) [reply lib/backend/dynamo/dynamodbbk.go:393]: The logic will work because we are relying on the `len(re.lastEvaluatedKey) == 0` check to verify if we scanned all elements thought settings the limit based on `limitRemaining = limit - len(result.records)` can introduce additional backend calls. For example if the limit was set to `1` and there are 1000 expired items the `getRecords` function will be invoked x1000 times. I have reverted this change.

File: lib/backend/test/suite.go
Comment (smallinsky) [reply lib/backend/test/suite.go:573]: I have found 3 tests failing from DynamoDB suite.  the DynamoDB item uses time.Now clock to check if item was expired thus calling `clock.Advance` doesn't enforce item to expired. 

Though after a second thought i have aligned time DynamoDB item logic to leverage the `b.Clock`

File: lib/backend/test/suite.go
Comment (greedy52) [inline lib/backend/test/suite.go:596]:
```suggestion
	// Fill the backend with events that total size is greater than 1MB (65KB * 20 > 1MB).
```

**Review by greedy52 — APPROVED** [review]

File: lib/backend/dynamo/dynamodbbk.go
Comment (r0mant) [inline lib/backend/dynamo/dynamodbbk.go:394]: So the issue was that we would just always enter this loop upon hitting the 1MB limit and never attempted to fetch another one? lgtm if so but please add a comment with explanation and wait for @fspmarshall's review as well.

**Review by fspmarshall — APPROVED** [review]
Fix for the 1mb issue looks good.  It seems like we're just mishandling limits here generally tho.

The backend should return at most `limit`, and should default to using `backend.DefaultRangeLimit` if no limit is specified.  I think we should change `getAllRecords` to set the limit to `backend.DefaultRangeLimit` if no limit is specified.  Also, instead of passing in the same limit for each subsequent call to `getRecords`, we should pass in `limit - len(result.records)`, so that we get at most *exactly* `limit` worth of items.

If you have the time, can you make those changes here b4 merging?

**Comment by smallinsky** [comment]
@fspmarshall 
> we should pass in limit - len(result.records), so that we get at most exactly limit worth of items.

This approach will have main drawback. Namely the limit is a applied before filtering in dynamoDB case so if  there are 1000x elements that doesn't match filtering rule and there is only one element that matches it will means tha in case of limit set to 1 the dynamoDB API in worst case will be called 1001x times. 

Regarding: 
> The backend should return at most limit, and should default to using backend.DefaultRangeLimit if no limit is specified

Agree, I will change this.

**Review by r0mant — APPROVED** [review]

**Pull request merged**
